opt: plan generic index joins

mgartner opened the pull request:
#### opt: refactor GenerateParameterizedJoin

GenerateParameterizedJoin has been refactored so that it constructs an
InnerJoin and Project in the custom function, and not in the optgen
replace pattern. This will make some upcoming changes simpler.

Release note: None

#### opt: plan generic index joins

The GenerateParameterizedJoin rule can now plan index joins instead of
inner joins that are later converted to lookup joins. Index joins are
more efficient than lookup joins because they are optimized for a
narrower set of capabilities and there is a vectorized implementation
for them. Index joins are currently only planned when all PK columns are
constrained to placeholders. This covers the most common cases of PK
lookups but this restriction can be loosened in the future (see the
added TODOs).

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140388)
<!-- Reviewable:end -->

mgartner commented:
Mixed results with `oltp_read_write` microbenchmark:

```
name                                        old time/op    new time/op    delta
Sysbench/SQL/1node_local/oltp_read_write-8    4.28ms ± 0%    4.33ms ± 0%  +1.10%  (p=0.000 n=10+10)

name                                        old alloc/op   new alloc/op   delta
Sysbench/SQL/1node_local/oltp_read_write-8    1.25MB ± 0%    1.22MB ± 0%  -2.62%  (p=0.000 n=10+10)

name                                        old allocs/op  new allocs/op  delta
Sysbench/SQL/1node_local/oltp_read_write-8     6.22k ± 0%     6.07k ± 0%  -2.45%  (p=0.000 n=10+10)
```

mgartner commented:
This optimization only applies to 2 `UPDATE` and 1 `DELETE` statement in sysbench, not any of the read queries. So the improvement shows up more clearly in `oltp_write_only`:

```
name                                        old time/op    new time/op    delta
Sysbench/SQL/1node_local/oltp_write_only-8    1.57ms ± 0%    1.53ms ± 0%  -2.62%  (p=0.000 n=9+10)

name                                        old alloc/op   new alloc/op   delta
Sysbench/SQL/1node_local/oltp_write_only-8     427kB ± 0%     393kB ± 0%  -8.04%  (p=0.000 n=10+8)

name                                        old allocs/op  new allocs/op  delta
Sysbench/SQL/1node_local/oltp_write_only-8     2.56k ± 0%     2.41k ± 0%  -6.08%  (p=0.000 n=8+10)
```

Next I'll run a full sysbench test to measure the throughput improvement.

DrewKimball reviewed (commented):
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140388)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/props/logical.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-OICF89SBJLvdFAKybN0:-OICF89SBJLvdFAKybN1:bdha9u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b492982c3856e9e6a5da2d921333f9d3cc267c/pkg/sql/opt/props/logical.go#L47)):*
> ```Go
> 	// GenerateParameterizedIndexJoin exploration rule has been applied to an
> 	// expression in the group. If set, GenerateParameterizedJoin does not fire.
> 	GeneratedParameterizedIndexJoin
> ```

I'm skeptical of this pattern where we set a logical property to avoid both rules firing and depend on their order. If possible I think it'd be better to find a way to combine them into one rule, or else make the index-join change somewhere else. For example, what if we made the `GenerateLookupJoins` rules create an index join as an optimization? Alternatively, what if we had a rule that converts a PK-lookup-join into an index-join?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/props/logical.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-OICF89SBJLvdFAKybN0:-OIC_v3A0JatNQyXmdS1:b-xsf3yt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b492982c3856e9e6a5da2d921333f9d3cc267c/pkg/sql/opt/props/logical.go#L47)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'm skeptical of this pattern where we set a logical property to avoid both rules firing and depend on their order. If possible I think it'd be better to find a way to combine them into one rule, or else make the index-join change somewhere else. For example, what if we made the `GenerateLookupJoins` rules create an index join as an optimization? Alternatively, what if we had a rule that converts a PK-lookup-join into an index-join?
</blockquote></details>

I agree, this is a bit of an abuse of what `AvailableRuleProps` was designed for. I liked this though because it allowed the rules to be separate which I thought would make both easier to understand. There's no technical reason the rules can't be combined. I'll do that instead.

I also thought about performing the transformation in GenerateLookupJoins, but that function is already fairly complicated and I don't think there is any cases outside of a parameterized lookup join that will use that transformation, so it didn't seem worth it.


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: refactor GenerateParameterizedJoin

GenerateParameterizedJoin has been refactored so that it constructs an
InnerJoin and Project in the custom function, and not in the optgen
replace pattern. This will make some upcoming changes simpler.

Release note: None

mgartner committed:
opt: plan generic index joins

The GenerateParameterizedJoin rule can now plan index joins instead of
inner joins that are later converted to lookup joins. Index joins are
more efficient than lookup joins because they are optimized for a
narrower set of capabilities and there is a vectorized implementation
for them. Index joins are currently only planned when all PK columns are
constrained to placeholders. This covers the most common cases of PK
lookups but this restriction can be loosened in the future (see the
added TODOs).

Release note: None

mgartner reviewed (commented):
I didn't measure any improvement to throughput in a 5-iteration sysbench test. Average throughput was actually down 1.5%:

### Before

```
23263 qps = (23320 + 23005 + 22900 + 23732 + 23359) / 5
```

### After

```
22905 qps = (22951 + 23078 + 22805 + 23174 + 22519) / 5
```

Like I said above, these index join will only be planned for the write queries, which I believe are 25% of the queries in the workload, so the benefit is limited. I think the variability we see with the workload might explain the regression I measured. The benefits are logical in theory and the`BenchmarkSysbench/oltp_write_only` benchmarks above are enough to justify this change, in my opinion.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-:-OIGJYNt4xtrcRAvER5c:b-4udvlw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/opt/props/logical.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-OICF89SBJLvdFAKybN0:-OIGJUVW6xjUMGLPzWKy:b-7hwbg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4b492982c3856e9e6a5da2d921333f9d3cc267c/pkg/sql/opt/props/logical.go#L47)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I agree, this is a bit of an abuse of what `AvailableRuleProps` was designed for. I liked this though because it allowed the rules to be separate which I thought would make both easier to understand. There's no technical reason the rules can't be combined. I'll do that instead.

I also thought about performing the transformation in GenerateLookupJoins, but that function is already fairly complicated and I don't think there is any cases outside of a parameterized lookup join that will use that transformation, so it didn't seem worth it.
</blockquote></details>

I think combining into a single rule worked well. PTAL.


<!-- Sent from Reviewable.io -->

mgartner commented:
More good results:

```
name                                  old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-8    5.01ms ± 1%    4.95ms ± 0%  -1.11%  (p=0.000 n=10+10)

name                                  old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-8    2.11MB ± 1%    2.08MB ± 0%  -1.72%  (p=0.000 n=10+9)

name                                  old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-8     9.94k ± 0%     9.77k ± 0%  -1.73%  (p=0.000 n=10+10)
```

DrewKimball reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 6 files at r2, 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-:-OIIIpHI-ymqN-JHryp1:b-8ewgme)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/xform/generic_funcs.go` line 121 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-OIIHC0R4Z1pTh-bTEd4:-OIIHC0R4Z1pTh-bTEd5:b1pxjz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9ff1762b3f1a2846b8e01b17891c6226c1117bc/pkg/sql/opt/xform/generic_funcs.go#L121)):*
> ```Go
> // see the GenerateParameterizedJoin rule for more details. This function only
> // succeeds if the given filters have placeholders or stable expressions.
> func (c *CustomFuncs) generateParameterizedInnerJoin(
> ```

Is there any potential for passing the newly-constructed Values + ScanPrivate into `generateLookupJoinsImpl`? At first glance, it actually seems pretty straightforward.


<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+

mgartner commented:
*[`pkg/sql/opt/xform/generic_funcs.go` line 121 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140388#-OIIHC0R4Z1pTh-bTEd4:-OISVa6oE217Pg446rB2:b-lqq5xr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9ff1762b3f1a2846b8e01b17891c6226c1117bc/pkg/sql/opt/xform/generic_funcs.go#L121)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is there any potential for passing the newly-constructed Values + ScanPrivate into `generateLookupJoinsImpl`? At first glance, it actually seems pretty straightforward.
</blockquote></details>

Yes, good idea. I'll try that in a follow-up PR.


<!-- Sent from Reviewable.io -->

mgartner commented:
build failure:

```
*
* ERROR: a SQL panic has occurred while executing the following statement:
* WITH tab_json AS (SELECT crdb_internal.pb_to_json('desc', descriptor)->'table' AS t FROM system.descriptor WHERE id = $1::REGCLASS), columns_json AS (SELECT json_array_elements(t->'columns') AS c FROM tab_json), columns AS (SELECT (c->>'id')::INT8 AS col_id, IF((c->'inaccessible')::BOOL, c->>'computeExpr', quote_ident(c->>'name')) AS expr FROM columns_json), indexes_json AS (SELECT json_array_elements(t->'indexes') AS idx FROM tab_json UNION ALL SELECT t->'primaryIndex' FROM tab_json), unique_indexes AS (SELECT idx->'name' AS name, json_array_elements(idx->'keyColumnIds')::STRING::INT8 AS col_id FROM indexes_json WHERE (idx->'unique')::BOOL), index_exprs AS (SELECT name, expr FROM unique_indexes AS idx INNER JOIN columns AS c ON idx.col_id = c.col_id) SELECT ('(' || array_to_string(array_agg(expr), ', ')) || ')' AS final_expr FROM index_exprs WHERE expr != 'rowid' GROUP BY name
*
*
* ERROR: a panic has occurred!
* invalid datum type given: REGCLASS, expected INT8
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	pkg/sql/conn_executor.go:1008
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	pkg/sql/colexecerror/error.go:137
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	pkg/sql/colexecerror/error.go:137
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.DatumToEncDatum
*   | 	pkg/sql/rowenc/encoded_datum.go:186
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next
*   | 	pkg/sql/plan_node_to_row_source.go:227
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next
*   | 	pkg/sql/colexec/columnarizer.go:248
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
*   | 	pkg/sql/colexec/invariants_checker.go:88
*   | github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColIndexJoin).next
*   | 	pkg/sql/colfetcher/index_join.go:336
*   | github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColIndexJoin).Next
*   | 	pkg/sql/colfetcher/index_join.go:162
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
*   | 	pkg/sql/colexec/invariants_checker.go:88
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*CancelChecker).Next
*   | 	pkg/sql/colexec/colexecutils/cancel_checker.go:54
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*vectorTypeEnforcer).Next
*   | 	pkg/sql/colexec/colexecutils/operator.go:152
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.constBytesOp.Next
*   | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecbase/const.eg.go:208
```

DrewKimball commented:
```
FROM system.descriptor
WHERE id = $1::REGCLASS
```
`id` is an integer column. Is the rule missing a check for the placeholder type? Or maybe it should be adding casts?

mgartner commented:
I added a check that the placeholder and PK column types match, and a test for this case. https://github.com/cockroachdb/cockroach/pull/140388/files#diff-1fd461e33b4da5b0d3d39efe826e4261df532d18fbc6e4d3d8d02b857914b792R84-R86

mgartner commented:
Drew and I discussed how this PR abuses index joins, which are typically only performed on PK values that have already been produced by a secondary index. 

Three things Drew pointed out specifically about this change:

> Yeah, that makes sense. I think we should probably document this behavior for the execution operators, just to prevent future changes from breaking this behavior.

> Thinking about this more, in order to make this change we'll also have to update the IndexJoin logical properties to reflect the fact that it can filter rows. Which seems a little unfortunate - so maybe we should set a field on the private to indicate whether looked-up rows are guaranteed to exist?

> There's also at least one rule PushLimitIntoIndexJoin which will have to be updated.

In theory, all of the perf benefits of this PR should be achievable by execbuilding generic lookup joins as scans, which was attempted several months ago in #136938. I'm going to revive and polish that draft and run some benchmarks to prove that. If it shows a similar improvement, I think it is likely the better solution.

michae2 commented:
> > Thinking about this more, in order to make this change we'll also have to update the IndexJoin logical properties to reflect the fact that it can filter rows. Which seems a little unfortunate - so maybe we should set a field on the private to indicate whether looked-up rows are guaranteed to exist?
> 
> > There's also at least one rule PushLimitIntoIndexJoin which will have to be updated.

I don't know if it's helpful, but we had to handle a similar problem for index joins with skip-locked behavior in https://github.com/cockroachdb/cockroach/pull/85720. Maybe something from there applies?
